### DISCUSSION TITLE
Increase hitwindow cutoff for relax mod

### DISCUSSION THREAD

**Discussion by cihe13375** [ppy/osu#13665] [Ideas]
Hi~ idk if this is another "not a issue", but I post it here anyway

As far as I know, hitwindow works differently for relax mod and normal play (in osu!std), as roughly illustrated below:
![图片](https://user-images.githubusercontent.com/33959010/123475038-6fa6b200-d62d-11eb-892d-489aae8886c3.png)
To summarize, in relax mod, there's a "no trigger" region in the hitwindow, so if you "hit" (more precisely, cursor leave the note) too early, judgement won't be triggered (in stable the note will "vibrate" and in lazer there's no response at all) and in many cases it finally cause a miss (since your cursor may have moved to next notes). Such cutoff is not very pleasing imo (since it make hitwindow not symmetric), but I understand why it's necessary.
From my tests, the hitwindow cutoff in stable is ~10ms (without DT, unrelated to OD). This can be confirmed by checking hit error statistics (that "- xxx ms - xxx ms avg" thing) after a play, or watching the hiterror bar during gameplay. Somehow this is even harsher than OD10(19.5ms), but that's it.
![图片](https://user-images.githubusercontent.com/33959010/123477329-771b8a80-d630-11eb-9bb2-5f15a0f67c2d.png)
However, in lazer, the hitwindow cutoff is like ~2ms (unrelated to OD), which is _way shorter_ than stable. Again this can be confirmed by looking at timing distribution or hiterror bar: (these 2 screenshots are from same map and I can make sure the cursor always reach hitobject early enough during gameplay)
![图片](https://user-images.githubusercontent.com/33959010/123477869-438d3000-d631-11eb-8a2f-419d4957be99.png)
As a result, it's much more possible to miss in lazer than is stable when playing with relax (e.g. if cursor reached the note at -25ms and leave at -5ms, it would be count as 300 in stable, but miss in lazer). And imo cutoff of 2ms is way harsher than reasonable since (I suck at this game but still) I think most players can't feel time difference of 2ms, and multiple other factors (like timing accuracy, input latency, frame rate, etc.) can have larger flunctations than that. Also such low cutoff means the hitwindow is heavily biased, so players (at least me) may adjust audio offset to compensate that, which... well I believe this game should not be played in that way.

My suggestion is making the hitwindow as symmetry as possible, as it's really the most reasonable choice for a rhythm game (you may argue that relax mod is not rhythm game, but since judgements tend to trigger _as early as possible_ in relax, imo we should make the "first half" of hitwindow larger anyway). I would recommend making the cutoff at the first edge of 100 and 300 (e.g. if 300 hitwindow is -25~25ms, then the cutoff is -25ms), and all "hit early" will be count as 300 as long as they're triggered. This will make 300 hitwindow symmetric while not causing extra 100s under current mechanism. 100 and 50 hitwindow are still biased but imo it's fine since getting acc is easier in relax and 100/50 don't occur often.

If the suggestion above won't happen (in a short time), I think at least the hitwindow cutoff in Classic mod should match stable, since it's what Classic mod meant to do.

**Comment by peppy** [comment]
How would you suggest this works? You've written a lot of text, but I think you may have missed the *reasoning* this is the case: Relax hits objects in time. Hitting early would make the audio feedback horrible.

**Reply by cihe13375** [reply]
imo the "timepoint of hit" (which is same as "timepoint to play hitsound" in normal play) is not really well-defined in relax; because instead of hitting the note (press the key) at a certain timepoint, in relax notes are hit in a _period_ (cursor reach the note ~ cursor leave the note, from a player's perspective). So we can't really have a "correct" timepoint to play hitsound.
If our goal is making audio feedback more natural, I think we may just play the hitsound as close to 0ms as possible:
(1) hitsound is at 0ms if cursor is on the note at that time.
(2) if cursor reaches the note after 0ms (and not causing miss), then hitsound is played as soon as cursor reaches;
(3) if cursor leaves the note before 0ms (and not in the "no trigger" region), then hitsound is played as soon as cursor leaves (and the note disappears).
Also judgement can be shown at the same time of hitsound played.

btw infering from your statement, hitsound should usually be played at ~-10ms (the "hitwindow cutoff") in relax in stable, but personally I didn't feel any uncomfort from that, so at least it's not a big issue to match stable behavior imo (I can't represent all players though, of course).

**Reply by peppy** [reply]
Very much disagree with this. Playing hitsounds early sounds wrong; hitting objects early because the mouse left them sounds wrong. I think this is all out of scope of this mod.

**Reply by PercyDan54** [reply]
Agree with the " Playing hitsounds early sounds wrong" part, but lazer has relax_leniency = 3, which is too small compared to stable.
I belive relax shouldn't be able to get early 100s
